test/python/pulse/test_block.py

```diff
@@ -749,3 +750,227 @@ def test_parametrized_context(self):
         ref_sched = ref_sched.insert(90, pulse.Delay(10, self.d0))
 
         self.assertScheduleEqual(block, ref_sched)
+
+
+class TestBlockFilter(BaseTestBlock):
+    """Test ScheduleBlock filtering methods."""
+
+    def test_filter_channels(self):
+        """Test filtering over channels."""
+        # test a non-nested block
+        blk = pulse.ScheduleBlock()
+        blk.append(pulse.Play(self.test_waveform0, self.d0))
+        blk.append(pulse.Delay(10, self.d0))
+        blk.append(pulse.Play(self.test_waveform1, self.d1))
+
+        filtered_blk = blk.filter(channels=[self.d0])
+        self.assertEqual(len(filtered_blk.channels), 1)
+        self.assertTrue(self.d0 in filtered_blk.channels)
+        ref_blk = pulse.ScheduleBlock()
+        ref_blk.append(pulse.Play(self.test_waveform0, self.d0))
+        ref_blk.append(pulse.Delay(10, self.d0))
+        self.assertEqual(filtered_blk, ref_blk)
+
+        filtered_blk = blk.filter(channels=[self.d1])
+        self.assertEqual(len(filtered_blk.channels), 1)
+        self.assertTrue(self.d1 in filtered_blk.channels)
+        ref_blk = pulse.ScheduleBlock()
+        ref_blk.append(pulse.Play(self.test_waveform1, self.d1))
+        self.assertEqual(filtered_blk, ref_blk)
+
+        filtered_blk = blk.filter(channels=[self.d0, self.d1])
+        self.assertEqual(len(filtered_blk.channels), 2)
+        for ch in [self.d0, self.d1]:
+            self.assertTrue(ch in filtered_blk.channels)
+        self.assertEqual(filtered_blk, blk)
+
+        # test a nested block
+        c0 = pulse.ControlChannel(0)
```

Comment:
Please move this to another function, e.g. `test_filter_channel_nested_block`. Testing multiple logics in a single test case sometimes makes debug harder. Single test for single functionality is always better :)

Reply:
Done! 😃  👉 https://github.com/Qiskit/qiskit-terra/pull/9971/commits/2f0ad68f94ee728abdf07e3d542cfd9f20ffbef3